docs/man/mclog.1.in
.Dd March 2, 2026
.Dt mclog 1
.Os mcplex.net
.Sh NAME
.Nm mclog
.Nd view log messages from multicast or files
.Sh SYNOPSIS
.Nm
.Op Fl c Ar configFile
.Op Fl F Ar filterExpression
.Op Fl d
.Op Ar files...
.Sh DESCRIPTION
.Nm
may be used to view log messages multicasted from instances of
.Xr mclogd 8 .  It may also be used to view binary log files that
were written by
.Xr mclogd 8 .
.Pp
The following command line options are available:
.Pp
.Bl -tag -width indent
.It Fl c Ar configFile
Specify the path to the
.Xr mclogd 8
configuration file.  The default is @prefix@/etc/mclogd.cfg
.It Fl F Ar filterExpression
Only display messages which match the given \fIfilterExpression\fR.
.It Fl d
Enable debugging messages on stderr.
.It Ar files...
If present,
.Nm
will read messages from the given files.  If not present,
.Nm
will read messages that were multicasted by
.Xr mclogd 8 .
.El
.Sh FILES
.Nm
uses libCredence for encryption and authentication.  Hence it
requires credence key files.
.Bl -tag -width indent
.It Pa ~/.credence/id_ed25519
The user's credence private key file, created with
.Xr credence 1 .  This file should be owned by the user and have
permissions 0600.  It must contain the private part of an Ed25519
RSA key pair.
.It Pa ~/.credence/id_ed25519.pub
The user's credence public key file, created with
.Xr credence 1 .  This file should be owned by the user and have
permissions 0600.  It must contain the public part of an Ed25519 key pair.
.It Pa ~/.credence/known_keys
The user's credence known services file.  This file must contain the public
keys of
.Xr mclogd 8
from the
.Xr mclogd 8
hosts being queried (one per line).  It is used to authenticate instances
of
.Xr mclogd 8 .
.El
.Sh SEE ALSO
.Xr mclogd 8 ,
.Xr credence 1
.Sh COPYRIGHT
Copyright (c) 2024 Daniel W. McRobb
.Mt dwm@mcplex.net
.Pp



ChangeLog
2026-03-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as mclog-0.0.14

	* configure, configure.ac: Update required version of libDwm,
	libDwmCredence and dwmgmk.

2026-03-23  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as mclog-0.0.13

	* README.md: initial README markdown for github

	* classes/src/DwmMclogLoopbackSender.cc: Use MCLOGD_DEFAULT_PORT
	instead of 3737.

2026-03-21  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* configure, configure.ac, docs/man: manpage updates.

2026-03-20  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* etc/mclogd.cfg.example: Change multicast port to 3737.

	* doc/man: Work on manpages.

	* classes/include/DwmMclogConfig.hh: Set
	LoopbackConfig::listenIpv4 to true by default.

2026-03-19  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* apps/mclogd/mclogd.cc: Use MCLOGD_DEFAULT_CONFIG_PATH.

	* tagged as mclog-0.0.12

2026-03-19  dwm  <dwm@ria.mcplex.net>

	* etc/freebsd/rc.d/mclogd: Minor cleanup: I don't need precmd.

	* etc/freebsd/+POST_INSTALL.in: Don't try to restart mclogd, just
	emit a reminder.

	* apps/mclogd/mclogd.cc: Use atexit() to remove our PID file.

2026-03-17  dwm  <dwm@ria.mcplex.net>

	* tagged as mclog-0.0.11

	* etc/freebsd/rc.d/mclogd: Add explicit mclogd_start to see if it
	fixes the restart during installation.

2026-03-13  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as mclog-0.0.10

	* etc/mclogd.cfg.example: Add size information for logs.

	* tagged as mclog-0.0.9

	* classes/include/DwmMclogConfig.hh,
	  classes/include/DwmMclogLogFile.hh,
	  classes/src/DwmMclogConfigLex.lex,
	  classes/src/DwmMclogConfigParse.y,
	  classes/src/DwmMclogLogFile.cc,
	  classes/src/DwmMclogLogFiles.cc,
	  classes/tests/TestConfig.cc,
	  classes/tests/inputs/TestConfig1.cfg:
	Add 'size' configuration for log files.  This sets a maximum
	size for the log file before it is rolled over.

2026-03-10  dwm  <dwm@nomnom.mcplex.net>

	* tagged as mclog-0.0.8




apps/mclog/mclog.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2025
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  @file mclog.cc
//!  @author Daniel W. McRobb
//!  @brief NOT YET DOCUMENTED
//---------------------------------------------------------------------------

extern "C" {
  #include <unistd.h>
}

#include <cstdlib>
#include <filesystem>
#include <fstream>

#include "DwmMclogConfig.hh"
#include "DwmMclogLogger.hh"
#include "DwmMclogMulticastReceiver.hh"
#include "DwmMclogMessageFilterDriver.hh"
#include "DwmMclogSettings.hh"

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
class MySink
  : public Dwm::Mclog::MessageSink
{
public:
  MySink()
      : _filter(nullptr)
  {
  }

  void SetFilter(const std::string & expr)
  {
    _filter = make_unique<Dwm::Mclog::MessageFilterDriver>(expr);
  }
  
  bool Process(const Dwm::Mclog::Message & msg) override
  {
    if (! _filter) {
      std::cout << msg << std::flush;
    }
    else {
      bool  result;
      if (_filter->parse(&msg, result) && result) {
        std::cout << msg << std::flush;
      }
    }
    return true;
  }

private:
  std::unique_ptr<Dwm::Mclog::MessageFilterDriver>  _filter;
};

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
void ProcessBZ2File(const char *filename,
                    std::shared_ptr<Dwm::Mclog::MessageFilterDriver> filter)
{
  BZFILE  *bzf = BZ2_bzopen(filename, "rb");
  if (bzf) {
    Dwm::Mclog::Message  msg;
    bool                 filterResult;
    while (msg.BZRead(bzf) > 0) {
      if ((! filter) || (filter->parse(&msg, filterResult) && filterResult)) {
        std::cout << msg;
      }
    }
    BZ2_bzclose(bzf);
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
void ProcessGZFile(const char *filename,
                   std::shared_ptr<Dwm::Mclog::MessageFilterDriver> filter)
{
  gzFile  gzf = gzopen(filename, "rb");
  if (gzf) {
    Dwm::Mclog::Message  msg;
    bool                 filterResult;
    while (msg.Read(gzf) > 0) {
      if ((! filter) || (filter->parse(&msg, filterResult) && filterResult)) {
        std::cout << msg;
      }
    }
    gzclose(gzf);
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
void ProcessIstream(std::istream & is,
                    std::shared_ptr<Dwm::Mclog::MessageFilterDriver> filter)
{
  Dwm::Mclog::Message  msg;
  bool                 filterResult;
  while (msg.Read(is)) {
    if ((! filter) || (filter->parse(&msg, filterResult) && filterResult)) {
      std::cout << msg;
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
void ProcessBinaryFile(const char *filename,
                       std::shared_ptr<Dwm::Mclog::MessageFilterDriver> filter)
{
  std::ifstream  is(filename);
  if (is) {
    ProcessIstream(is, filter);
    is.close();
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
void ProcessFile(const std::filesystem::path & path,
                 std::shared_ptr<Dwm::Mclog::MessageFilterDriver> filter)
{
  if (path.string() == "-") {
    ProcessIstream(std::cin, filter);
  }
  else {
    auto  ext = path.extension();
    if (ext.string() == ".bz2") {
      ProcessBZ2File(path.string().c_str(), filter);
    }
    else if (ext.string() == ".gz") {
      ProcessGZFile(path.string().c_str(), filter);
    }
    else {
      ProcessBinaryFile(path.string().c_str(), filter);
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void Usage(const char *argv0)
{
  std::cerr << "usage: " << argv0
            << " [-c configFile] [-d] [-F filterExpression] [files...]\n";
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  bool         debug = false;
  std::string  configFile{MCLOGD_DEFAULT_CONFIG_PATH};
  std::string  filtexpr;
  int          optChar;
  while ((optChar = getopt(argc, argv, "c:dF:")) != -1) {
    switch (optChar) {
      case 'c':
        configFile = optarg;
        break;
      case 'd':
        debug = true;
        break;
      case 'F':
        filtexpr = optarg;
        break;
      default:
        Usage(argv[0]);
        exit(1);
        break;
    }
  }

  std::shared_ptr<Dwm::Mclog::MessageFilterDriver>  filter{nullptr};
  if (! filtexpr.empty()) {
    filter = make_shared<Dwm::Mclog::MessageFilterDriver>(filtexpr);
  }

  if (optind < argc) {
    //  We've been given filenames on command line.  Process them and
    //  then exit.
    for (int i = optind; i < argc; ++i) {
      ProcessFile(argv[i], filter);
    }
    exit(0);
  }

  Dwm::Mclog::OstreamSink  cerrSink(std::cerr);
  if (debug) {
    Dwm::Mclog::logger.Open(Dwm::Mclog::Facility::user, {&cerrSink}, "mclog");
  }
  
  std::string  keyDir("~/.credence");
  Dwm::Mclog::MulticastReceiver  mcastRecv;
  Dwm::Mclog::Config  config;
  if (config.Parse(configFile)) {
    config.service.keyDirectory = keyDir;
    if (mcastRecv.Open(config)) {
      MySink  mysink;
      if (! filtexpr.empty()) {
        mysink.SetFilter(filtexpr);
      }
      mcastRecv.AddSink(&mysink);
      for (;;) {
        sleep(60);
      }
    }
  }
  
  
  return 0;
}

          
          
